Handle cursor area with custom writer

lammel opened the pull request:
With this merge request using a custom writer is now possible by using a instanced cursor area.
Moving the cursor will be restricted to the size of the area. Multiline is also supported.

Marked as draft, as tests have only been conducted under Linux and probably some discussion is needed.

MarvinJWendt commented:
Hi, awesome PR, thanks! 

I am unable to compile it under Linux currently:
![image](https://github.com/atomicgo/cursor/assets/31022056/a158db1a-3177-425c-98ac-870d929e4ef2)

I like the idea, and I think implementation wise, there is not much to discuss (looks good to me).

MarvinJWendt reviewed (commented):
I assume that you do not have a Windows environment? If so, what would be the most appropriate way for you to implement it for Windows? Let me know if you need any help with that :)

lammel commented on cursor.go:
```diff
@@ -5,66 +5,148 @@ package cursor
+// Cursor displays content which can be updated on the fly.
+// You can use this to create live output, charts, dropdowns, etc.
+type Cursor struct {
+	writer io.Writer
```
Changed to `Writer` interface and moved global cursor routines to `utils.go` for now.

The global helpers will ensure compatibility with current interface (I even added Top for symmetry).
I did not fully understand the purpose of the autoheight handling of the cursor helpers, but the behaviour be unchanged.

Hope that resolves your remarks.

MarvinJWendt replied:
Huh, did you push that change? 😅 
Still looks to be `io.Writer`.

lammel replied:
> `io.Writer` -> `Writer`

Overlooked windows, should be correct now (committed 3 days)

MarvinJWendt commented on area.go:
```diff
@@ -2,50 +2,161 @@ package cursor
 // NewArea returns a new Area.
-func NewArea() Area {
-	return Area{
+func NewArea() *Area {
```
Is there a reason that this is now returning a pointer?

lammel replied:
Reverted in latest commit, no change required. 
Can't even say why I did that in the first place ^^

MarvinJWendt commented on area.go:
```diff
@@ -2,50 +2,161 @@ package cursor
-// WithWriter sets a custom writer for the Area.
-func (area Area) WithWriter(writer Writer) Area {
+// WithWriter sets the custom writer
+func (area *Area) WithWriter(writer io.Writer) *Area {
```
`WithXxx` methods must not accept a pointer receiver. If you have area `a` and you do `b := a.WithWriter(...)` you would change the writer of `a` and `b`. 

Also, we need to accept the internal `Writer` interface here, instead of `io.Writer`.

lammel replied:
Corrected

MarvinJWendt commented on area.go:
```diff
@@ -2,50 +2,161 @@ package cursor
 	fmt.Fprintln(area.writer, content)
+	// Detect height of cursor area
+	area.height = strings.Count(content, "\n")
+	if strings.HasSuffix(content, "\n") || runtime.GOOS == "windows" {
```
Could you please add a comment here, why windows has a different behavior?

lammel replied:
Changed to not require that special handling at all and not add an unneeded newline (use `Fprint` instead of `Fprintln`). Adjustments might be required for pterm `InteractiveTextInputPrinter` concerning cursor movement (which also has issues with the current version)

MarvinJWendt commented on cursor.go:
```diff
@@ -5,66 +5,84 @@ package cursor
+
+// WithWriter allows for any arbitrary Writer to be used
+// for cursor movement abstracted.
+func (c *Cursor) WithWriter(w io.Writer) *Cursor {
```
```suggestion
func (c *Cursor) WithWriter(w Writer) *Cursor {
```

MarvinJWendt commented on cursor_windows.go:
```diff
@@ -1,46 +1,53 @@
+
+// WithWriter allows for any arbitrary Writer to be used
+// for cursor movement abstracted.
+func (c *Cursor) WithWriter(w io.Writer) *Cursor {
```
```suggestion
func (c *Cursor) WithWriter(w Writer) *Cursor {
```

MarvinJWendt commented on cursor_windows.go:
```diff
@@ -1,46 +1,53 @@
+// for cursor movement abstracted.
+func (c *Cursor) WithWriter(w io.Writer) *Cursor {
+	if w != nil {
+		c.writer = w.(Writer)
```
```suggestion
		c.writer = w
```

MarvinJWendt reviewed (changes requested):
Hi, I have left some comments.

Just for reference:

We need to use the internal `Writer` interface everywhere, instead of `io.Writer`. Windows handles cursor movement via syscalls, which means we need a file descriptor. If you would do this:
```go
func main() {
	var buffer bytes.Buffer
	a := cursor.NewArea().WithWriter(&buffer)
	a.Update("Hello World!")
}
```
We would get this panic on Windows:
![image](https://github.com/atomicgo/cursor/assets/31022056/46c1f419-c4c3-491f-a76f-648d1e732d2e)

We have to do this for every operating system, as otherwise, Linux / macOS users would be able to do the example above, which would then panic on Windows. By accepting the internal `Writer` interface, we make sure that there has to be a file descriptor.

The only downside to this is, that the area cannot write to a buffer, but that isn't really relevant, as there is no cursor to move in a buffer anyway.

The rest looks good to me :)

// PS:

Sorry for the delay, I was on vacation.

lammel commented:
> Hi, I have left some comments.
> 
> Just for reference:
> 
> We need to use the internal `Writer` interface everywhere, instead of `io.Writer`. Windows handles cursor movement via syscalls, which means we need a file descriptor.

Done. pterm also requires a PR to make use of `cursor.WithWriter` which will be handled by a followup PR.

> The rest looks good to me :)

Great!

> Sorry for the delay, I was on vacation.

Nice, will do the same next week. So in case we haven't merged then, we might need some more days then 🆗

lammel commented:
I ran the area tests on my virtualized Windows 10 VM with good results so far. 
Behaves exactly the same on Linux and Windows.

But would be good if you give it a try. Thanks for the support here.

lammel commented:
Just fixed the compile error on windows, caused by a missing commit.
Please let the workflow run again (PR for CI lint is also available now)

MarvinJWendt commented:
Hi @lammel, there are some linting errors, and they (finally) are reported correctly. You can see them here: https://github.com/atomicgo/cursor/pull/17/files

Or if you have docker installed, you can use this command locally: `docker run --rm -v $(pwd):/app -w /app golangci/golangci-lint:v1.53.1 golangci-lint run`.

Btw: You don't need to update the `README.md` files in AtomicGo packages. They are generated automatically when the PR is pushed to master 😉

lammel commented:
> Hi @lammel, there are some linting errors, and they (finally) are reported correctly. You can see them here: https://github.com/atomicgo/cursor/pull/17/files
> 
> Or if you have docker installed, you can use this command locally: `docker run --rm -v $(pwd):/app -w /app golangci/golangci-lint:v1.53.1 golangci-lint run`.

I'm using a locally installed golangci-lint now.
The rules seem a little too harsh (exhaustive struct, cuddled blocks), but I've fixed those linter warnings now.
 
> Btw: You don't need to update the `README.md` files in AtomicGo packages. They are generated automatically when the PR is pushed to master wink

Thanks for noting, good to know and saves some time :)

lammel commented:
> Hi @lammel, sorry for the late response, I was on vacation. I have implemented an important fix that caused the area to be messy on Windows. Could you please merge the changes of #20 into this PR? Thanks! (On Update Windows should use `Fprint` and everything else `Fprintln` and print line by line)
> 
> ```go
> // Update overwrites the content of the Area.
> func (area *Area) Update(content string) {
> 	oldWriter := target
> 
> 	SetTarget(area.writer) // Temporary set the target to the Area's writer so we can use the cursor functions
> 	area.Clear()
> 	lines := strings.Split(content, "\n")
> 	fmt.Fprintln(area.writer, strings.Repeat("\n", len(lines)-1)) // This appends space if the terminal is at the bottom
> 	Up(len(lines))
> 	SetTarget(oldWriter) // Reset the target to the old writer
> 
> 	// Workaround for buggy behavior on Windows
> 	if runtime.GOOS == "windows" {
> 		for _, line := range lines {
> 			fmt.Fprint(area.writer, line)
> 			StartOfLineDown(1)
> 		}
> 	} else {
> 		for _, line := range lines {
> 			fmt.Fprintln(area.writer, line)
> 		}
> 	}
> 
> 	height = 0
> 	area.height = len(strings.Split(content, "\n"))
> }
> ```

Actually I removed the newline for linux on purpose, as single line updates would otherwise leak into the next line, although this is not required and requires more cursor movement.
So I'd like to keep the minimal cursor movement and fix the cursor/area usage in pterm.
Or do you see compatibility issues for other software using cursor?

MarvinJWendt commented:
I think that should work, the linux implementation didn't had any bugs. For Windows the `StartOfLineDown(1)` part is important. Otherwise, it would result in this: 

https://user-images.githubusercontent.com/31022056/254358832-a62121e1-bbb9-4529-94d5-7470b1d198dc.mp4

lammel commented:
> I think that should work, the linux implementation didn't had any bugs. For Windows the `StartOfLineDown(1)` part is important. Otherwise, it would result in this:
> 
>  idea64_uDtapqB6WY.mp4 
> Also, I couldn't reproduce the bug on Windows consistently. The select and multiselect printers were affected all the time, but I couldn't really isolate the problem. Sometimes the Windows terminal is just a little weird.

As this is a pterm test that might also cause unexpected behaviour I've added some tests under `_examples` here now to reproduce the issues.

I could reproduce 1 windows issue and one generic issue you were facing. 
For windows the cursor package now ensures `\r\n` is used for `area.Update()` which fixes the cursor not moving to the start of the line. This is done in a platform dependent helper in `area_windows.go` now. Furthermore this problem will probably be resolved by future Windows updates (hopefully).

The second issue was reproduced by `_example/movement/main.go` which showed a jumpy behavior of the area in case content length differs (e.g. when searching in pterm`s `interactive_select` example)

Both of these issues are resolved now. The only remaining problem seems to be moving the cursor `Up()` out of bounds of the area, which causes the next update to garble output.

Your tests with pterm should run a lot smoother now.

lammel commented:
> > Both of these issues are resolved now. The only remaining problem seems to be moving the cursor Up() out of bounds of the area, which causes the next update to garble output.
> 
> Could this be fixed by preventing to call `Up()` when the cursor is in line 1 of the area?

Should be, just ran out of time yesterday, I'll look into it soon. 
Once merged this should allow pterm to have a quite reliable cursor area for the interactive inputs which also handle single line inputs.

lammel commented:
This PR is now ready to merge. All added examples here and for pterm work as expected on Linux and Windows.
For pterm please be sure to test with PR https://github.com/pterm/pterm/pull/529

lammel commented:
Ah one more note... This PR may break current behavior depending on the use of the cursor area. You might consider that when merging/releasing.

MarvinJWendt commented:
> After making the linter happy again, the above issues are fixed now.
> 
> So only the pterms multiline input has issues with the custom cursor positioning, The y-axis for cursorPosY is a little awkward, I'd suggest to correct it to 0x0 being the upper left (not lower left) of the terminal area (already that way in pterm). Any objections?

Just to clarify, you mean the position in the `Area` right? If so, no objections from my side :)

MarvinJWendt reviewed (approved):
LGTM! Thanks for contributing 🚀

PS: Your examples look awesome!

lammel commented:
> Quick addition, I have invited you to this repo as collaborator (which gives you some more rights). If you want, you can accept the invitation here: https://github.com/atomicgo/cursor/invitations.
> 
> Don't feel pressured by this to do more, but take it as appreciation :)

Thanks! I feel a little pressured ;-)
But I'm already looking into adding missing pieces to pterm (cursor here was just a requirement to get Stderr working there), so why not help out a little more 🆗
